### ISSUE TITLE
Pin down the pin system semantics

### ISSUE THREAD

**Issue by dbuenzli** [ocaml/opam#2156]
See discussions in http://lists.ocaml.org/pipermail/opam-devel/2015-May/001068.html and https://github.com/ocaml/opam/issues/2141

The semantics is currently confusing and the default mode of operation dangerous in the sense that it's too dependent on the current state of your file system, which makes it easy to pick up unwanted state when you deal with many pins over a long period of time, as you need to remember the exact checkout state of your pins whenever you do an `opam upgrade -u`. Besides currently vcs-pins are completely inconsistent in the sense that the behaviour is very different on a `pin add -kgit` depending on whether `REPO` is an URL or a local repo.

The pin system should be changed so that the following mental model is preserved whether `REPO` is an URL or a local repository which is not the case at the moment,  `opam -kgit add pin REPO`:
1. opam makes its own private `git clone REPO`
2. opam runs `git pull` on this clone on opam `update`

W.r.t. to the current behaviour it means that adding a new `git`-pin no longer tracks `HEAD`, but the currently checkout-out branch in the repo (the value of `git symbolic-ref HEAD`). Recovering the old behaviour can be achieved by specifying `REPO#HEAD` on the command line. The `opam pin list` dashboard should also be made more explicit so that's it's easier to see what is going on by printing `$PKG git $BRANCH $REPO` rather than the hard to parse `PKG git $REPO#$BRANCH`.

In order to support workflows where you may want to test for a change in the working directory of the pin without having to commit on the tracked branch. The `-w/--working-dir` option is added to `opam upgrade/install/reinstall`. In this case the semantics is as follows, rather than looking at the branch information of the pin, opam picks up the on-disk state of git-tracked files currently checked out in the working directory of the pin's git repo and uses this to install the package. This state is kept on further opam updates until the branch of the pin changes again or a new `-w` upgrade is requested.

I would personnally say the mixed-mode pins as they exist now should not be needed with such a system and can be killed. Simple path pinning should simply rsync. More precisely the use-case for mixed-mode was:
1. Being able to test changes in a pin without having to commit and without having the build pick up build artefacts (which can occur with simple path-pinning). This use case is well supported with `-w`. 
2. Use of `(re)install` that don't allow you to see if changes were picked up. This is rather a command misuse. By using `opam upgrade` you will always know if changes are being picked up or not.

**Comment by dbuenzli** [comment]
If an `upgrade` fails, the package is no longer installed. A further `upgrade` after fixing will lead to this:

```
> opam upgrade bos

=-=- Synchronising pinned packages =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=  🐫 
[bos] /Users/dbuenzli/sync/repos/bos#master updated
[NOTE] bos is not installed, ignored.

Already up-to-date.
```

It would be nice if this could behave the same way `opam reinstall` behaves if the package is not installed so that one can always use `opam upgrade` to have a nice vcs pin experience.

```
> opam reinstall bos

=-=- Synchronising pinned packages =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=  🐫 
[bos] /Users/dbuenzli/sync/repos/bos#master already up-to-date
[WARNING] bos is not installed.
Install ? [Y/n] y
```

**Comment by AltGr** [comment]
On your latest comment, 20cde4e86d83b45df09d896cc8af9e7620d1822f will probably come as a relief: it allows `opam install -u` to install-or-upgrade, which you can run for building your pinned package without having to think twice about its current state.

**Comment by dbuenzli** [comment]
Thanks but is it still possible to have `opam upgrade` behave as `opam reinstall` if the package is not installed ? Having a single constant command to use without an option will minimize errors for the pin workflow. I'm now always using `opam upgrade` and it works very well except when your package fails to install in which case you have to use `opam install` once afterwards. Using `opam install` there's always the chance that I will forget about `-u` and it is also describes less what I'm actually doing.

**Comment by AltGr** [comment]
Yes indeed.

**Comment by dbuenzli** [comment]
There is also another issue I often run into. When you develop for the next release you do: 

```
opam pin add -kgit PKG PKG-repo#master
```

You then make you release, it gets into the opam repository you do an `opam update` and then you do a 

```
opam pin remove PKG
```

In this case it will reinstall the version of `PKG` that was there when you pinned `PKG`, is there any good reason why it doesn't try to install the latest version provided the installed constraints allow it ?

The current behaviour for me often either leads to downgrades which I need to reupgrade and/or breakage, followed by a manuall reinstall of many packages. This is annoying if your package is a highly used dependency.

**Comment by dbuenzli** [comment]
> In this case it will reinstall the version of PKG that was there when you pinned PKG, is there any good reason why it doesn't try to install the latest version provided the installed constraints allow it ?

I have been doing quite a few releases recently and from a package releaser perspective this is *very* annoying (and is also to people you ask to beta-test through pin).

**Comment by AltGr** [comment]
As a temporary workaround, until this is fixed, I would advise to `opam pin remove --no-action`, then `opam upgrade`.

**Comment by dbuenzli** [comment]
Also the following message on `opam upgrade` could be improved as it is a bit ambiguous: it's never clear to me whether you updated the pin, possibly realizing there are no new commits, or if there are new commits. 
```
=-=- Synchronising pinned packages =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=  🐫 
[fmt] /Users/dbuenzli/sync/repos/fmt#master updated
```
I would rather use `changed`.

**Comment by samoht** [comment]
Is there any plan to make `-k git` be Git mode? I have been bitten by this a lot recently. If you want to keep the "mixed" mode, maybe add `-k mixed`?

**Comment by AltGr** [comment]
I tend to agree with Daniel's proposal. Just didn't get to the top of my list yet.
For the interface to be consistent, you must consider that mixed _is_ git mode, it's just that if you don't specify a branch (or ref) it synchronises the "Working tree" rather than the "HEAD".

As a quick fix, I could make `opam pin add` select the current branch automatically (maybe only when git is explicitely selected ?) ; but I'd prefer to get to something more sensible directly than add to the confusion with a half-baked workaround. This is difficult because expecations vary and both mixed or pure git lead to confusion (confusion the other way around was the reason `mixed` was created, let's not forget and loop eternally).

**Comment by dbuenzli** [comment]
> This is difficult because expecations vary and both mixed or pure git lead to confusion (confusion the other way around was the reason mixed was created, let's not forget and loop eternally).

We are not looping and I'd say you are forgetting. AFAIR `mixed` mode was a misguided attempt at solving 1. and 2 (see report above). People were in trouble with 1. with path pinning and `oasis` because builds would pickup outdated `setup.data` which would result in inconsistent builds.

There's no reason to be confused if a pin behaves consistently whether you pin a local git repo or a remote one; if people are being confused by this I'd rather say that their brain is.

**Comment by AltGr** [comment]
I didn't say we were, just that we need to be careful not to.

**Comment by Drup** [comment]
(Just want to weight and say that I actually like the mixed mode semantics ...)

**Comment by dbuenzli** [comment]
> (Just want to weight in and say that I actually like the mixed mode semantics ...)

What do you think about the git + `-w` workflow ?

**Comment by Drup** [comment]
I like the fact that the information on the semantic of the pin is provided when doing pin, not by upgrade/install/reinstall. 

I'm pretty sure I'm going to forget about the `-w` everytime I want it. On the other hand, if I want the non-mixed semantics, I will make that decision at pin time, and it's going to stay that way. I just think this is a pin-time decision, not an upgrade-time one. 

However, I do agree that the mixed semantics may not be the best default, I'm not sure which one is best.

**Comment by Drup** [comment]
Note: I use the mixed mode 80% of the time, for stuff that I actually dev, path mode for long-to-build things (typically with opam 2.0, ocaml switches) and the git-to-branch mode for various specific things (typically, when I want my opam universe to use a branch while doing stuff in another branch, usually for things that I do work on very actively).

**Comment by dbuenzli** [comment]
> I'm pretty sure I'm going to forget about the -w everytime I want it.

Sure, but the cost of you forgetting this is going to be quite small (nothing changed) w.r.t. to an upgrade that picks up an unwanted repository working directory state and potentially screws your switch (if the package is used by many others).

> I just think this is a pin-time decision, not an upgrade-time one.

I'd disagree on that one, my own personal experience is that this doesn't work well in practice beyond a few pins unless you are particularly well disciplined in your `upgrade`s and constantly remember the repo checkout state of all your pins (or do this only for short lived periods). Of course YMMV. 

Regarding the default I really think `-k git` should not change its behaviour depending on whether the git repo is local or remote, so it should really not be mixed mode. Note what I'm not talking about  `opam pin add`'s default, which I care a little bit less I'm now used to `-kgit` my pins, but I'm annoyed that I have to do `-kgit` + `repo#master` on local pins while I don't have anything special to do on remote ones, I constantly have to redo my local pins twice because I forget.

**Comment by hannesm** [comment]
I completely agree to @dbuenzli that `-k git` should not change its behaviour depending on whether the git repo is local or remote.

In the proposal, please keep in mind that `git symbolic-ref HEAD` is not always well-defined (there might be `fatal: ref HEAD is not a symbolic ref` if you have a detached `HEAD`).  I rarely end up in such a state while developing my own libraries, but do regularly while debugging, esp. if debugging on a remote machine.

**Comment by dbuenzli** [comment]
> In the proposal, please keep in mind that git symbolic-ref HEAD is not always well-defined (there might be fatal: ref HEAD is not a symbolic ref if you have a detached HEAD)

I don't know what happens if you try to `git clone` a repo that is in this state, but the result should be equivalent to this (including error if it is an error). I really want to be able to think of `git` pins in terms of the mental model I gave in my proposal (see first point 1) and 2)).

**Comment by Drup** [comment]
> my own personal experience is that this doesn't work well in practice beyond a few pins unless you are particularly well disciplined in your upgrades

I must say, I'm not even sure what exactly are the dangers you mentioned. I never add any issue with mixed pinning picking unexpected state from my working directory. All the files that are "state" (build and configure stuff) are safely gitignored and are never picked in mixed mode, and I rarely leave a local working directly in complete disarray. 

Furthermore, I tend to isolate projects in various switches and pins are separated (so that I will not mess with jsoo while working on mirage stuff, for example). The main switch (typically, the one with the tooling) doesn't contain any pin.

> I completely agree to @dbuenzli that -k git should not change its behaviour depending on whether the git repo is local or remote.

If that is the issue, I wouldn't mind having `-kgit` with a consistent behavior and a new kind `-kmixed` with the current one. Even better if we get an env variable to specify the default pinning mode. :)

**Comment by dbuenzli** [comment]
Le mardi, 5 juillet 2016 à 15:43, Gabriel Radanne a écrit :

> > my own personal experience is that this doesn't work well in practice beyond a few pins unless you are particularly well disciplined in your upgrades
>  
> I must say, I'm not even sure what exactly are the dangers you mentioned. I never add any issue with mixed pinning picking unexpected state from my working directory. All the files that are "state" (build and configure stuff) are safely gitignored and are never picked in mixed mode, and I rarely leave a local working directly in complete disarray.
> Furthermore, I tend to isolate projects in various switches and pins are separated (so that I will not mess with jsoo while working on mirage stuff, for example). The main switch (typically, the one with the tooling) doesn't contain any pin.

That's precisely what I mention: this all requires a lot of attention and discipline which I personally don't have. A usable system shouldn't require too much discipline from its user in order to function in a satisfying way. The current defaults yield brittle setups by default unless you are very careful.  

Daniel

**Comment by hannesm** [comment]
@dbuenzli `git clone` on a `detached HEAD` repo clones into the same detached HEAD state; a `git pull` later will complain `not on a branch`. just FYI

**Comment by dbuenzli** [comment]
So nothing was done on this front in opam v2 ?

**Comment by AltGr** [comment]
I was expecting stuff to come up with the new local switches and `--inplace-build`, as defining more streamlined workflows will probably incur tweaks in the pinning system. So I'm hoping to fit this in the broader story.

Also, my feedback may not be representative (the ones giving it aren't the one with the least knowledge of the internals), but I have not heard other complaints about mixed mode. The two conflicting points here are "not deceiving expectations", while being as simple as possible. Mixed-mode was designed for the first but, I agree, adds too much mental complexity to the model.

**Comment by dbuenzli** [comment]
> The two conflicting points here are "not deceiving expectations", while being as simple as possible. Mixed-mode was designed for the first but, I agree, adds too much mental complexity to the model.

Mixed-mode was certainly not designed for the first. It's sad that I have to repeat points so much but again there should be no difference if I git-pin a remote directory or a local one, *that* is deceiving expectations. 

I think the usability problems have been made enough at that point, @samoht even says he was also caught by this more than once. I thought it would be quite evident that shared mutable state (i.e. mixed mode) doesn't scale.

**Comment by AltGr** [comment]
> Mixed-mode was certainly not designed for the first. It's sad that I have to repeat points so much but again there should be no difference if I git-pin a remote directory or a local one, that is deceiving expectations. 

No need to repeat, I got your point; if you remember, in 1.1, ppl where doing `opam switch foo ./` and getting puzzled that opam didn't pick up their changes because it created a git pin -- the idea of mixed mode came up to avoid that. I'm not saying it's the best behaviour and it shouldn't be changed (that's not even my opinion), but it at least managed to follow _some_ expectations.

Also, it seems you took my message the wrong way, I was actually saying that something should and would be done about this, just that I wanted to make it fit in the bigger picture first. So really, I got your point here.

**Comment by dbuenzli** [comment]
> No need to repeat, I got your point; 

Apparently not: 

> if you remember, in 1.1, ppl where doing opam switch foo ./ and getting puzzled that opam didn't pick up their changes because it created a git pin -- the idea of mixed mode came up to avoid that

Please reread the whole thing the problem is not to change that default, the problem is to change the default when you ask for a git pin.

**Comment by AltGr** [comment]
Alright, I'm _finally_ getting back to this.

Summarising the grudges above:
- _upgrade vs reinstall annoyances_: fixed, now both can work, not depending on the current state. As an unrelated bonus, failing to recompile should'nt remove the previously installed version anymore, so there is that too
- _behaviour after `opam unpin`_: the default is now to upgrade
- _unclear update message_: should be fixed too
- _Don't select mixed mode by default_: done, git is now purely git (not pushed yet).
- _Select the current branch when pinning_: on it, shouldn't be long

I've also implemented `--working-dir`, but for now it reuses the code of mixed mode, i.e. it synchronises the current state of the version-controlled files only. I don't think that's the right choice, as it gives puzzling results when files are added or removed. Another idea is to sync everything but what's on the VCS ignore list (`.gitignore`) ; but that could also have corner cases, e.g. with the not uncommon outdated or inaccurate `.gitignore`, so a raw sync is still better IMHO: the worst case is a missed `make clean` before `opam install --working-dir`, and at least that is easy to diagnose and fix.

I am not sure about the proposed persisting state for `--working-dir`, either, that feels like adding another mode besides pinning. In its current state, `--working-dir` only lives until the package is next updated, which can happen by:
 1. `opam install|reinstall|upgrade`ing it, when the package is specified on the command-line only
 2. `opam update`ing the package explicitely
 3. Running a global `opam update`

In particular, a rebuild caused by a change to a dependency will keep the state cached in opam. I think the cases 1. and 2. above are fine, too, since in these cases you would probably repeat the `--working-dir` (note that there should be an option to avoid that auto-update though. I thought there was one already but it seems not). The last case may be more problematic, so maybe global updates should skip packages in a dirty state with a message.

Thanks again for all the feedback on this, and I'd be glad to get more opinions on the changes presented above.

**Comment by AltGr** [comment]
Another issue, actually, is that currently, opam fetches the new source from upstream whenever the package is going to be (re)installed. I think it should keep its cached version except in the cases above (`update`, or when the package is listed on the CL)

**Comment by AltGr** [comment]
#2770 should provide a general fix for all of the above. No more mixed, but a warning with uncommitted changes, `--working-dir` that lives until the next explicit update (and is printed about on global update).

`--working-dir` still synchronises the version-controlled files only, though.
